File: .chezmoitemplates/codex/extract-project-trust.py
from __future__ import annotations

import pathlib
import sys


def is_header(line: str) -> bool:
    stripped = line.strip()
    return stripped.startswith("[") and stripped.endswith("]")


def is_project_header(line: str) -> bool:
    stripped = line.strip()
    return stripped == "[projects]" or stripped.startswith("[projects.")


def main() -> int:
    if len(sys.argv) != 2:
        return 0

    path = pathlib.Path(sys.argv[1]).expanduser()
    if not path.exists():
        return 0

    lines = path.read_text(encoding="utf-8").splitlines()
    blocks: list[str] = []
    current: list[str] | None = None

    for line in lines:
        if is_header(line):
            if current:
                blocks.append("\n".join(current).rstrip())
                current = None
            if is_project_header(line):
                current = [line]
            continue

        if current is not None:
            current.append(line)

    if current:
        blocks.append("\n".join(current).rstrip())

    output = "\n\n".join(block for block in blocks if block.strip())
    if output:
        sys.stdout.write(output)
        if not output.endswith("\n"):
            sys.stdout.write("\n")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: .chezmoitemplates/codex/extract-project-trust.js
"use strict";

function isHeader(line) {
  const stripped = line.trim();
  return stripped.startsWith("[") && stripped.endsWith("]");
}

function isProjectHeader(line) {
  const stripped = line.trim();
  return stripped === "[projects]" || stripped.startsWith("[projects.");
}

function extractProjectTrust(text) {
  const lines = text.split(/\r\n|[\n\r\v\f\x1c-\x1e\x85\u2028\u2029]/u);
  const blocks = [];
  let current = null;

  for (const line of lines) {
    if (isHeader(line)) {
      if (current !== null) {
        blocks.push(current.join("\n").trimEnd());
        current = null;
      }
      if (isProjectHeader(line)) current = [line];
      continue;
    }

    if (current !== null) current.push(line);
  }

  if (current !== null) blocks.push(current.join("\n").trimEnd());
  const output = blocks.filter((block) => block.trim()).join("\n\n");
  return output ? `${output}\n` : "";
}

function main(argv) {
  if (argv.length !== 1) {
    process.stderr.write(
      "usage: extract-project-trust.js <existing-config-text>\n",
    );
    return 2;
  }

  try {
    process.stdout.write(extractProjectTrust(argv[0]));
    return 0;
  } catch (error) {
    process.stderr.write(`extract-project-trust.js: ${error.stack}\n`);
    return 1;
  }
}

module.exports = { extractProjectTrust };

if (require.main === module) process.exitCode = main(process.argv.slice(2));
